pkg/audit/writer.go

```diff
@@ -0,0 +1,178 @@
+package audit
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	lumberjack "gopkg.in/natefinch/lumberjack.v2"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+)
+
+const (
+	contentTypeJSON = "application/json"
+)
+
+const (
+	levelMetadata = iota
+	levelRequest
+	levelRequestResponse
+)
+
+var (
+	auditLevel = map[int]string{
+		levelMetadata:        "MetadataLevel",
+		levelRequest:         "RequestLevel",
+		levelRequestResponse: "RequestResponseLevel",
+	}
+
+	bodyMethods = map[string]bool{
+		http.MethodPut:  true,
+		http.MethodPost: true,
+	}
+)
+
+const (
+	requestReceived  = "RequestReceived"
+	responseComplete = "ResponseComplete"
+)
+
+type Log struct {
+	AuditID        k8stypes.UID `json:"auditID,omitempty"`
+	RequestURI     string       `json:"requestURI,omitempty"`
+	RequestBody    interface{}  `json:"requestBody,omitempty"`
+	ResponseBody   interface{}  `json:"responseBody,omitempty"`
+	ResponseStatus string       `json:"responseStatus,omitempty"`
+	SourceIPs      []string     `json:"sourceIPs,omitempty"`
+	User           *userInfo    `json:"user,omitempty"`
+	UserAgent      string       `json:"userAgent,omitempty"`
+	Verb           string       `json:"verb,omitempty"`
+	Stage          string       `json:"stage,omitempty"`
+	StageTimestamp string       `json:"stageTimestamp,omitempty"`
+}
+
+type userInfo struct {
+	Name  string   `json:"name,omitempty"`
+	Group []string `json:"group,omitempty"`
+}
+
+type LogWriter struct {
+	Level  int
+	Output *lumberjack.Logger
+}
+
+func NewLogWriter(path string, level, limitSize int) *LogWriter {
+	if path == "" {
+		return nil
+	}
+	var maxBackup, maxSize int
+	if limitSize < 100 {
+		maxBackup = 1
+		maxSize = limitSize
+	} else {
+		maxBackup = limitSize / 100
+		maxSize = 100
+	}
+	return &LogWriter{
+		Level: level,
+		Output: &lumberjack.Logger{
+			Filename:   path,
+			MaxBackups: maxBackup,
+			MaxSize:    maxSize,
+		},
+	}
+}
+
+func (a *LogWriter) LogRequest(req *http.Request, auditID k8stypes.UID, authed bool, contentType, user string, group []string) error {
+	al := &Log{
```

Comment:
do we need to check the level at the beginning of the function, similar to how we do for LogResponse? 

```
if a.Level < levelRequestResponse {
		return nil
	}
```

Reply:
LogRequest func will log metadata and log request body, if the level is higher than the requestBody level, will log the request body, otherwise will only log the metadata, so we don't need to add the check at the beginning because we still need to log metadata.